src/main/java/it/gov/pagopa/payment/service/QRCodePaymentServiceImpl.java

```diff
@@ -0,0 +1,79 @@
+import it.gov.pagopa.payment.model.TransactionInProgress;
+import it.gov.pagopa.payment.repository.RewardRuleRepository;
+import it.gov.pagopa.payment.repository.TransactionInProgressRepository;
+import it.gov.pagopa.payment.utils.PerformanceLogUtils;
+import it.gov.pagopa.payment.utils.TrxCodeGenUtil;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class QRCodePaymentServiceImpl implements QRCodePaymentService {
+
+  private final TransactionInProgress2TransactionCreatedMapper
+      transactionInProgress2TransactionCreatedMapper;
+  private final TransactionCreationRequest2TransactionInProgressMapper
+      transactionCreationRequest2TransactionInProgressMapper;
+  private final RewardRuleRepository rewardRuleRepository;
+  private final TransactionInProgressRepository transactionInProgressRepository;
+  private final TrxCodeGenUtil trxCodeGenUtil;
+
+  public QRCodePaymentServiceImpl(
+      TransactionInProgress2TransactionCreatedMapper transactionInProgress2TransactionCreatedMapper,
+      TransactionCreationRequest2TransactionInProgressMapper
+          transactionCreationRequest2TransactionInProgressMapper,
+      RewardRuleRepository rewardRuleRepository,
+      TransactionInProgressRepository transactionInProgressRepository,
+      TrxCodeGenUtil trxCodeGenUtil) {
+    this.transactionInProgress2TransactionCreatedMapper =
+        transactionInProgress2TransactionCreatedMapper;
+    this.transactionCreationRequest2TransactionInProgressMapper =
+        transactionCreationRequest2TransactionInProgressMapper;
+    this.rewardRuleRepository = rewardRuleRepository;
+    this.transactionInProgressRepository = transactionInProgressRepository;
+    this.trxCodeGenUtil = trxCodeGenUtil;
+  }
+
+  @Override
+  public TransactionCreated createTransaction(TransactionCreationRequest trxCreationRequest) {
+
+    Long startTime = System.currentTimeMillis();
+
+    if (!rewardRuleRepository.checkIfExists(trxCreationRequest.getInitiativeId())) {
+
+      log.error("Cannot find initiative with ID: [{}]", trxCreationRequest.getInitiativeId());
+
+      throw new ClientExceptionWithBody(
+          HttpStatus.NOT_FOUND,
+          "NOT FOUND",
+          "Cannot find initiative with ID: [%s]".formatted(trxCreationRequest.getInitiativeId()));
+    }
+
+    TransactionInProgress trx = transactionCreationRequest2TransactionInProgressMapper.apply(trxCreationRequest);
+
+    generateTrxCodeAndSave(trx);
+
+    PerformanceLogUtils.performanceLog("CREATE_TRANSACTION_QR_CODE", startTime);
+
+    return transactionInProgress2TransactionCreatedMapper.apply(trx);
+  }
+
+  private void generateTrxCodeAndSave(TransactionInProgress trx) {
+    boolean exists = true;
+    String trxCode = null;
+    while (exists) {
+      trxCode = trxCodeGenUtil.get();
+      exists = transactionInProgressRepository.existsByTrxCode(trxCode);
```

Comment:
let's test as first if it could works: transactionCode is not the _id, so maybe it will always be an insert and never a modified

we should test storing a trxCode and then calling this method verifying if it will create or update (we should setOnInsert the id field!)